flake-module.nix

```
@@ -0,0 +1,78 @@
+{ self, lib, flake-parts-lib, ... }:
+let
+  inherit (flake-parts-lib)
+    mkPerSystemOption;
+  inherit (lib)
+    mkOption
+    types;
+in
+{
+  options = {
+    perSystem = mkPerSystemOption
+      ({ config, self', inputs', pkgs, system, ... }: {
+        options.treefmt = mkOption {
+          description = ''
+            treefmt flake module options
+          '';
+          type = types.submodule {
+            options = {
+              config = mkOption {
+                description = "treefmt-nix configuration";
+                type = types.submoduleWith {
+                  modules = (import ./.).all-modules pkgs;
+                };
+              };
+              module = mkOption {
+                type = types.raw; # TODO: module type?
+                description = "Evaluated module of treefmt-nix";
+                default =
+                  (import ./.).evalModule pkgs config.treefmt.config;
+              };
+              programs = mkOption {
+                type = types.attrsOf types.package;
```

```suggestion
                type = types.listOf types.package;
```
A set of all the packages isn't it?

---

I used an attrset rather than a list - because, while the likes of `buildInputs` does take a list, in other places attrsets are used. 

For example, `haskell-flake`'s `buildTools` is an attrset of packages, because it allows the user to override the default build tools to disable some of them (by setting `{ foo = null; }`. 

```nix
          buildTools = hp: {
            treefmt = config.treefmt.wrapper;
          } // config.treefmt.programs;
```

https://github.com/srid/haskell-template/blob/1706660e7b02ece881a67c99692b70db5ae5c8db/flake.nix#L28-L30

---

I also figured it would be nice to "tag" each program with the formatter it is associated with it, so the caller can do any special processing that way. If we make this use `listOf` instead, then haskell-template will have to re-implement the logic.